### DISCUSSION TITLE
Implementing /health

### DISCUSSION THREAD

**Discussion by jpraet** [belgif/rest-guide#165] [General]
Could you share some experiences, and maybe documentation if you have it available, about the implementation of the /health endpoint in your REST APIs?

* Do you transitively call the /health endpoint on downstream dependencies of the API to calculate the availability status? 
  * Do these in turn call /health on their dependencies (which might be APIs of external suppliers), and so on?
  * Do you protect against circular dependencies?
* Do you cache the availability status? If so: on which level, and for how long?
* Do you expose the /health check to your clients, or is it only for internal use? Does it require an access token?

**Comment by pvdbosch** [comment]
> * Do you transitively call the /health endpoint on downstream dependencies of the API to calculate the availability status?
>   * Do these in turn call /health on their dependencies (which might be APIs of external suppliers), and so on?

Yes, recommended that backends implement it like this. Backend API may use server/framework health-check functionality (e.g. retrieve status from Spring Boot actuator's health check) that makes this easier.

>   * Do you protect against circular dependencies?

Circular dependency probably indicates a bad architecture; no generic detection. Backends could implement a workaround at design time if circular dependency is known.

> * Do you cache the availability status? If so: on which level, and for how long?
> * Do you expose the /health check to your clients, or is it only for internal use? Does it require an access token?

I'd have to check about this, but I believe access requires a specific access scope.

**Comment by jpraet** [comment]
> Circular dependency probably indicates a bad architecture; no generic detection.

Agreed. We do have at least one example though at CBSS: SSKM -> SsinInformationService -> SSKM.
Generic detection approach would be to pass a Cbss-Health-Visited HTTP header containing the upstream /health call chain.
When the health check component detects itself already present in that call chain, it returns UP.

**Comment by jpraet** [comment]
@VirginieHayot could you share your experiences about implementing these /health checks at EHealth?

There is some reluctance at our IT/System department about transitively calling /health on the downstream dependencies.
If we can have confirmation that SMALS and EHealth also implement their health checks that way, I think that can settle the argument.

**Reply by pvdbosch** [reply]
Some additions:
* Some transitive dependencies might not be used in the call chain for every type of incoming call, so could result in 'DEGRADED' rather than 'DOWN' status. But generally, I'd say if all of the main functionality of the API is entirely unavailable, DOWN is preferred. For instance, refData operations often have less dependencies and have more uptime, but aren't the main reason why the API is being used.
* The health check described in the guide is end-user oriented (i.e. is the service working OK?), not a technical one (is the deployed component working OK?). Kubernetes liveness probes, unlike health checks, shouldn't call external dependencies bc they would cause restarts of the container.

**Comment by VirginieHayot** [comment]
@jpraet 
- yes, we also do transitive calls (same as for our SOAP monitoring). It has been proven to be complicated to have that recursive logic into monitoring tools, with the help of cartography, to derive if service A is down that service B is also impacted. It makes it more transparent and non-ambiguous if a main transitive provider is down, that this API also becomes unusable. Not aware of any circular dependencies, we don't have explicit protection against those.
- we typically cache the result and have a small internal framework based on Spring to help hide the plumbing. Per default the scheduled recalculate is done every second, but with external systems we tend to loosen that timer to every minute. In any way, the downside of this, is that it's no longer real-time: we accept that monitoring calls show a slight stale status. The advantage is clearly that we buffer any monitoring calls, we are in charge of the pace with which a backend system is monitored and also protect it this way. Win-win.
- The /health operation can only be called with a token, and is protected by a separate scope and in all cases I know, with a Client Credentials OAuth flow. We do not offer this to external clients for now. Only our Supervision monitoring teams calls it.

**Comment by JDMKSZ** [comment]
> >   * Do these in turn call /health on their dependencies (which might be APIs of external suppliers), and so on?
> 
> Yes, recommended that backends implement it like this. Backend API may use server/framework health-check functionality (e.g. retrieve status from Spring Boot actuator's health check) that makes this easier.

@pvdbosch This is not explicitly mentioned in the style guide, shouldn't we make this more explicit? It might seem obvious (otherwise it's just a "/ping" and you can never have e.g. a status "degraded") but it's better to avoid misinterpretation.

**Reply by pvdbosch** [reply]
Agreed, we can add a note to the guide.

Maybe like this:

> The status should indicate the availability of the API's functionality to its clients. This status typically depends on the availability of multiple components used to provide the API's functionality (e.g. application container, database, external web service dependency, ...).
  It differs from a container liveness status, which only reflects the availability of a single component.

**Reply by JDMKSZ** [reply]
Looks good to me.

- I think this can be added to the [rule](https://www.belgif.be/specification/rest/api-guide/#rule-hlth-res)?
- "should" -> "SHOULD" (3rd word of the paragraph)
- Should we add that a "container liveness status should not be checked with the /health operation, instead another operation (e.g. /ping) should be implemented"?

**Reply by jpraet** [reply]
> Should we add that a "container liveness status should not be checked with the /health operation, instead another operation (e.g. /ping) should be implemented"?

I think liveness / readiness status checks are out of scope of the REST guide.
On Spring Boot, this is typically `/actuator/health/liveness` and `/actuator/health/readiness`.
With MicroProfile, this is typically `/health/live` and `/health/ready`.

**Reply by pvdbosch** [reply]
Created PR: https://github.com/belgif/rest-guide/pull/174

Bc REST guide should be readable by functional analysts and the rule is already quite large, I put the most of it in an attention point below the rule.

**Comment by jpraet** [comment]
How do you combine the health status of the transitive dependencies into a global status?

For DOWN, we can make a distinction between critical and non-critical dependencies:

* A critical dependency that is DOWN, renders the global status DOWN
* A non-critical dependency that is DOWN, renders the global status DEGRADED

But what if the transitive dependency is DEGRADED? DEGRADED means the component is available with reduced functionality.
Does that render the global status DEGRADED too?
Or do you also make some distinction there regarding how critical the dependency is?

The top-level service could:

* not be impacted at all by the degraded functionality of a dependency (desired global status UP)
* be degraded itself as a result of the degraded dependency (desired global status DEGRADED)
* require the degraded functionality of the dependency to function correctly (desired global status DOWN)

But implementing such logic would require knowledge and interpretation of the health details of dependencies.

How do Smals/eHealth/others handle this?

**Reply by pvdbosch** [reply]
There's no single approach within our organization, but I believe the correct way is indeed having specific API-specific logic to deduce the correct aggregated status. I haven't seen much use of the DEGRADED status yet.

For Spring Boot, a custom [StatusAggregator](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#actuator.endpoints.health.writing-custom-health-indicators) can be written, but it doesn't seem to be able to take into account context information. Maybe a custom CompositeHealthContributor could be added that takes into account context information, but (some of the) auto-configured health indicators may need to be disabled then.

A default StatusAggregator supporting DEGRADED could be added in a library, but it would have to make a sane default choice between one of the behaviors (probably component-level DEGRADED resulting in top-level DEGRADED?)
